### ISSUE TITLE
Disable Ribbon client-side load-balancing to avoid Zuul Timeouts when updating pods

### ISSUE THREAD

**Issue by hannesrohde** [spring-cloud/spring-cloud-kubernetes#180]
We use Kubernetes service discovery in combination with Zuul proxying in an API gateway we develop. The gateway provides authenticated access to a bunch of microservices inside a Gcloud project.

This works mostly fine, but we noticed that sometimes we would get a Zuul timeout when a backend service behind the gateway undergoes a rolling update and olds pods are replaced with new instances.

Closer examination revealed that the Ribbon client-side load-balancing employed by Zuul uses an instance of `org.springframework.cloud.kubernetes.ribbon.KubernetesServerList` per route to get list of service instances.
This list can contain outdated entries for instances that have recently been shut down. The situation improved when reducing `ribbon.ServerListRefreshInterval` to a low value (e.g. 1000ms), but the server list still lags behind the current state of the deployment.

If possible we would like to change that behavior and instead route requests using the dns-entry for a service instead of routing to individual instances.

Our approach now would be to provide a custom ribbon client configuration where the server list contains a single static entry like this:

```
public class CustomRibbonClientConfiguration {

    class SingleHostServerList extends AbstractServerList<Server> {

        private final List<Server> servers = new ArrayList<>(1);

        SingleHostServerList(IClientConfig config) {
            String serviceName = config.getClientName();
            servers.add(new Server(getDnsName(serviceName), 80));
        }

        @Override
        public void initWithNiwsConfig(IClientConfig clientConfig) {
        }

        @Override
        public List<Server> getInitialListOfServers() {
            return servers;
        }

        @Override
        public List<Server> getUpdatedListOfServers() {
            return servers;
        }
		
        private String getDnsName(String serviceName) {
            return "...";	// e.g. "my-svc.my-namespace.svc.cluster.local"
        }
		
    }

    @Bean
    public ServerList<Server> ribbonServerList(IClientConfig config) {
        return new SingleHostServerList(config);
    }

    @Bean
    public IPing ribbonPing() {
        return new NoOpPing();
    }
}
```
```
@Configuration
@RibbonClients(defaultConfiguration = CustomRibbonClientConfiguration.class)
public class RibbonConfiguration {
}
```
Is this approach feasible?
Are there any preferable ways to get rid of the Ribbon client-side load-balancing while keeping Zuul as reverse proxy?

Used versions:
spring-boot-starter-web:jar:1.5.12.RELEASE
spring-cloud-starter-zuul:jar:1.3.6.RELEASE
spring-cloud-starter-kubernetes:jar:0.2.0.RELEASE

**Comment by hannesrohde** [comment]
@spencergibb we would still like to discover routes to services dynamically, just not the single instances of the services.

**Comment by spencergibb** [comment]
I'm confused then. Your example above has a hard coded server.

**Comment by MattSenter** [comment]
I'm interested in this use case as well. My specific requirement is that I need to use Eureka in my local environment and Kube in my cloud environment. I need a way to keep my FeignClient as `@FeignClient(name="my-service")` and have that work for both scenarios. So, if I'm understanding OP, he wants to have a custom resolve for the service name. Expanding on OP's suggestion, I have a custom config:

```
@Configuration
@Profile("cloud")
public class CloudDiscoveryConfig {

  @RibbonClient(name = "my-service-1", configuration = CloudRibbonConfig.class)
  public static class MyService1Config {
  }

  @RibbonClient(name = "my-service-2", configuration = CloudRibbonConfig.class)
  public static class MyService2Config {
  }

}
```

...both of which reference my `CloudRibbonConfig` which is almost identical to OP's code above. It seems a bit heavy handed. It also seems weird because it's hammering server-side load balancing into a client-side load balancing configuration (i.e. client-side load balancer of 1). I'd be interested in hearing alternatives.

**Comment by hannesrohde** [comment]
@spencergibb sorry for the confusion, let me try again :)
We want to discover service names and then use host names *derived* from the service names for the routes, thus using the load-balancing provided by Kubernetes.

@MattSenter yeah, my workaround definitely abuses Ribbon and I would very much prefer a solution that gets rid of it at all instead of decapacitating it. Good to hear I'm not the only one scratching his head over this.

**Comment by spencergibb** [comment]
I see. Instead of a `@RibbonClient` per service, use `@RibbonClients(defaultConfiguration)`. Ribbon is meant to be customized.

**Comment by hannesrohde** [comment]
Our project uses the `CustomRibbonClientConfiguration` as `defaultConfiguration`, I've edited the initial post and added the missing @Configuration part.

I am fine with using that if there is no obvious better solution, I was just curious if defeating Ribbon in this way is the right approach.

**Comment by prafsoni** [comment]
+1 to this request. There should be a way to utilize load balancing baked in K8s.

**Comment by spring-issuemaster** [comment]
Closing due to age of the question. If you would like us to look at this issue, please comment and we will look at re-opening the issue.
